vispy/visuals/visual.py

```diff
@@ -347,34 +349,111 @@ def set_gl_state(self, preset=None, **kwargs):
 
         The arguments are forwarded to :func:`vispy.gloo.wrappers.set_state`.
 
+        This can also be used as a context manager that will revert the
+        gl_state on exit. When used as a context manager, this function is
+        designed to be constructed directly in the header of the `with`
+        statement to avoid confusion about what state will be restored on exit.
+
         Parameters
         ----------
         preset : str
             Preset to use.
         **kwargs : dict
             Keyword arguments.
         """
+        prev_gl_state = self._vshare.gl_state.copy()
+
         self._vshare.gl_state = kwargs
         self._vshare.gl_state['preset'] = preset
 
+        @contextmanager
+        def _context():
+            self._prev_gl_state.append(prev_gl_state)
+            try:
+                yield self._vshare.gl_state
+            finally:
+                self.pop_gl_state()
```

Comment:
Is it worth making this a separate method or external function (still `_` prefixed) so that the methods making a context manager can share it? You could also explicitly pass `prev_gl_state` (and `self` depending on if it works as a method or a separate function).

Reply:
Sure - updated in d58014a69b42d1af79c06d57d1eedc3f44b8ab92. For generality this loses the returned "current state" from the context manager but I think that's okay as it was inconsistent between `Visual` and `CompoundVisual` anyway.